### ISSUE TITLE
Why the polygon labels need to be shrunk?

### ISSUE THREAD

**Issue by SUZhaoyu** [argman/EAST#94]
As mentioned in the original [paper](https://arxiv.org/pdf/1704.03155v2.pdf)  (Session 3.3.1):

> The positive area of the quad- rangle on the score map is designed to be roughly a shrunk version of the original one.

And this is how they operate:

> For each edge ⟨pi, p(i mod 4)+1⟩, we shrink it by moving its two endpoints inward along the
edge by 0.3ri and 0.3r(i mod 4)+1 respectively.

But may I ask why this operation is needed?

**Comment by argman** [comment]
I dont know, maybe @zxytim can help

**Comment by SUZhaoyu** [comment]
Thanks a lot. @argman

**Comment by zxytim** [comment]
@SUZhaoyu The affinity of the boundary area of a text line is ambiguous: it could be mine, yours or even cannot tell. However, the belonging of the center area of a text line quite clear: it should be inside a text line. We choose to be less ambiguous, so as to deem only the center area as inside a text line.

**Comment by SUZhaoyu** [comment]
Thank you for your reply @zxytim . I created my own dataset and trained from scratch, and I found that  the inference results always intended to be a bit smaller than expectation and now I think I know why.

**Comment by weinman** [comment]
@SUZhaoyu and @zxytim,  Conversely, there also may be a benefit to setting the training mask to zero (meaning they would ignored as "don't care" values) in the regions of the rectangle that were shrunk away. This code currently leaves the training mask as ones in these outer regions of the rectangle, where the class label has been set to zero.

**Comment by JC1DA** [comment]
I have a question. In the paper, for RBOX, the author shows the rectangle covers the original box instead of the shrunk one. Do we really need to shrink it?

**Comment by weinman** [comment]
@JC1DA The predicted coordinates go out to the original box dimensions, but the model is only trained so that these predictions happen where they are most reliable (in the inner portion of the box).
My more recent experiments definitely confirm that setting the training mask (not the class labels) to zero in the shrunk away region gives far more stable and improved training results. The original EAST paper is silent on this point, but the code in this repo sets the shrunk away region to a negative [non-text] class label, which I believe is problematic.

**Comment by JC1DA** [comment]
@weinman thanks so much for your information. I think we can increase the R-value a bit to mitigate the problem? Also setting the training mask as you mentioned might help.

**Comment by weinman** [comment]
@JC1DA What is the problem you're trying to mitigate by increasing R (the proportion of the axis that is shrunk away)?

**Comment by JC1DA** [comment]
@weinman sometimes I have bounding boxes inside the text. Thinking a way to get rid of it.

**Comment by JC1DA** [comment]
@weinman BTW, Can I ask you this? After conv layer, we have a bigger receptive field. So after de-conv layer or upsampling, is the receptive field reduced?
Someone said this model had trouble with long text due to the limited receptive field. I'm wondering if it is true or not.

**Comment by YoadTew** [comment]
> @JC1DA The predicted coordinates go out to the original box dimensions, but the model is only trained so that these predictions happen where they are most reliable (in the inner portion of the box).
> My more recent experiments definitely confirm that setting the training mask (not the class labels) to zero in the shrunk away region gives far more stable and improved training results. The original EAST paper is silent on this point, but the code in this repo sets the shrunk away region to a negative [non-text] class label, which I believe is problematic.

@weinman Im trying to train the code on my own dataset, and I happened to see your comment, and it sounds like a good idea. Could you please share your own changed code, or point out which part of the code (which file and lines) I need to change in order to do it?
Thanks!

**Comment by weinman** [comment]
@YoadTew The input pipeline for the EAST code here is horrendously inefficient. I completely rewrote it and will publish it separately when it is ready. Meanwhile, the idea is to add some lines in icdar.generate_rbox that not only set the [ground truth to 1](https://github.com/argman/EAST/blob/dca414de39a3a4915a019c9a02c1832a31cdd0ca/icdar.py#L480). I keep a mask image zero except for where the original rectangle is:

```
        # Invariant: rect_mask all 0 before this
        cv2.fillPoly(rect_mask, rect.astype(np.int32)[np.newaxis,:,:], 2)
        cv2.fillPoly(rect_mask, shrunk_rect, 1)
```

Then I set the training mask to zero where rect_mask is 2 (inside rectangle) but not 1 (inside shrunk_rectangle):

```
        # Unlike argman/EAST, ignore the pixels in the ground truth rectangle
        # that were shrunk away, rather than treating them as "negative" labels
        clear_eroded = rect_mask==2 # eroded area

        # Also want to clear any new positive labels that are already labeled,
        # to avoid learning confusion about potentially conflicting outputs
        clear_intersection = np.logical_and( rect_mask!=0, 
                                             training_mask!=2 )

        # Set to train on shrunk_rect
        cv2.fillPoly(training_mask, shrunk_rect, 1) 


        to_clear = np.logical_or( clear_eroded, clear_intersection)
        training_mask[np.nonzero(to_clear)] = 0
```
Hopefully that gives you some idea of how you might proceed.

**Comment by alwc** [comment]
Hi professor @weinman, is `training_mask!=2` in `clear_intersection` a typo?

**Comment by weinman** [comment]
No, it's not a typo, but a poorly worded preface. Thanks for double checking.

The purpose of the `clear_intersection` is to ignore in training on any pixel that appears in two separate training rectangles. (Which rectangle would you rather the model be trained predict? A question I chose to avoid.)

Variable `rect_mask` indicates the locations of the current rectangle, thus locations where `rect_mask!=0` means the current rectangle is potentially to be included in training. The variable `training_mask` contains the accumulated categorization of every pixel as follows

 `0`. definitely excluded (either because it's part of the shrunk-away rectangle region or it belongs to more than one rectangle)
 `1`. definitely included (because it's within the shrunk-away rectangle and does not yet belong to more than one rectangle)
 `2`. unknown (not involved in any rectangles yet)

So if `training_mask!=2`, that means the pixel is either already excluded for some reason (`0`) or is already included as a positive (`1`). If that is the case *and* we have a new rectangle (indicated by `rect_mask!=0`) at the same location, we want to exclude it from the training; hence the conjunction defining `clear_intersection`.
